packages/xstate-wallet/src/browser-wallet.ts

```diff
@@ -0,0 +1,187 @@
+      logger.info('No incoming objectives');
+    } else {
+      await this.store.addObjective(payloadObjective);
+    }
+
+    // Store any new states
+    const payloadState = payload.signedStates?.[0];
+    if (!payloadState) {
+      logger.info('No incoming states');
+    } else {
+      await this.store.addState(payloadState);
+      const channelId = calculateChannelId(payloadState);
+      if (!this.registeredChannels.has(channelId)) {
+        this.chain.chainUpdatedFeed(channelId).subscribe({
+          next: chainInfo => this.onFundingUpdate(channelId, chainInfo)
+        });
+      }
+    }
+
+    // Crank objectives
+    for (const objective of this.store.objectives) {
+      switch (objective.type) {
+        case 'OpenChannel': {
+          response = await this.onOpenChannelObjective(objective.data.targetChannelId);
+          break;
+        }
+        default:
+          throw new Error('Objective not supported');
+      }
+    }
+    return response;
+  }
+
+  async onOpenChannelObjective(channelId: string): Promise<Message> {
+    const channel = await this.store.getEntry(channelId);
+    const pk = await this.store.getPrivateKey(await this.store.getAddress());
+    const depositInfo = await this.getDepositInfo(channelId);
+
+    const response = this.crankOpenChannelObjective(
+      channel,
+      this.channelFunding[channelId],
+      depositInfo,
+      pk
+    );
+    if (response.deposit) {
+      this.channelFunding[channelId] = {...this.channelFunding[channelId], status: 'DEPOSITED'};
+      await this.chain.deposit(
+        channel.channelId,
+        this.channelFunding[channelId].amountOnChain,
+        depositInfo.myDeposit
+      );
+    }
+    if (response.signedStates[0]) {
+      await this.store.addState(response.signedStates[0]);
+      this.onNewMessage(response);
+    }
+    return response;
+  }
+
+  crankOpenChannelObjective(
+    channel: ChannelStoreEntry,
+    channelFunding: Funding,
+    depositInfo: DepositInfo,
+    pk: string
+  ): Response {
+    const response: Response = {
+      objectives: [],
+      signedStates: []
+    };
+    const {latestState} = channel;
+    // Should we sign the prefund state?
+    if (latestState.turnNum === 0 && !channel.isSupportedByMe) {
+      const newState = channel.signAndAdd(latestState, pk);
+      response.signedStates = [newState];
+      return response;
+    }
+
+    // Should we deposit?
+    if (
+      BN.gte(channelFunding.amountOnChain, depositInfo.depositAt) &&
+      BN.lt(channelFunding.amountOnChain, depositInfo.totalAfterDeposit) &&
+      channelFunding.status !== 'DEPOSITED'
+    ) {
+      response.deposit = true;
+      return response;
+    }
+
+    // Should we sign the postfund state?
+    if (
+      BN.gte(channelFunding.amountOnChain, depositInfo.fundedAt) &&
+      latestState.turnNum === 3 &&
+      channel.latestSignedByMe.turnNum === 0
+    ) {
+      const newState = channel.signAndAdd(latestState, pk);
```

Comment:
How correct is this supposed to be? (It doesn't check that the postFS and the preFS have the same appData & outcome.)

Reply:
https://github.com/statechannels/statechannels/pull/3402#discussion_r603335387